protocol/sequencer-enforced-invariants.md

```diff
@@ -0,0 +1,119 @@
+- Sequencer-enforced invariants cannot create bugs in the rollup’s fault proof program that effect withdrawals.
+
+Sequencer-enforced invariants minimize complexity, yet still maintain rollup sovereignty. User sovereignty can also be maintained by allowing a fault or in(validity) proof of the delegated sequencer to [automatically trigger](https://x.com/sam___iamm/status/1798008895488827662) the SystemConfig to fall back to OP Stack defaults.
+
+### Related Problems
+
+MEV capture designs such as [Priority Is All You Need](https://www.paradigm.xyz/2024/06/priority-is-all-you-need) also use the sequencer role to enforce inclusion and ordering invariants and may require special handling of deposits. However, we note that sequencer-enforced invariants are more relevant for heavy and complex computation like exploit detection algorithms.
+
+# Proposed Solution
+
+## Hooks for configurable static analysis of L1 → L2 Messages
+
+Message validation hooks can be used within the bridging contracts to ensure that deposit messages will not violate sequencer-enforced invariants. These hooks are optional, turned off by default, and managed through the rollup’s SystemConfig.
+
+### Message Validation Hook Design
+
+Reference implementation: https://github.com/ethereum-optimism/optimism/pull/10791
+
+Two storage variables are added to the `SystemConfig` contract on L1 and, correspondingly, the `L1Block` contract on L2: `L1_MESSAGE_VALIDATOR_SLOT` and `L2_MESSAGE_VALIDATOR_SLOT`. The addresses at these storage slots, if they are non-zero, MUST adhere to the `IL1MessageValidator` and `IL2MessageValidator` interfaces respectively.
+
+```solidity
+interface IL1MessageValidator {
+		// Returns a boolean indicating if the message passes additional validation.
+    function validateMessage(
+        address _from,
+        address _to,
+        uint256 _mint,
+        uint256 _value,
+        uint64 _gasLimit,
+        bool _isCreation,
+        bytes memory _data
+    )
+        external
+        view
+        returns (bool);
+}
+```
+
+```solidity
+interface IL2MessageValidator {
+    // Returns a boolean indicating if the message passes additional validation.
+    function validateMessage(
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        bytes calldata _message
+    )
+        external
+        view
+        returns (bool);
+}
+```
+
+The `IL1MessageValidator` is static called within the `OptimismPortal` during the `_depositTransaction` function before the `TransactionDeposited` event is emitted. If the message validation fails, then the call reverts.
+
+The `IL2MessageValidator` is called within the `L2CrossDomainMessenger` during the `relayMessage` function before the external call of a given message. If message validation fails, then the message is added to the `failedMessages` mapping and the function returns. 
```

Comment:
You can get around this by withdrawing directly from the `L2ToL1MessagePasser`

Reply:
It's not clear to me why that is the case, but it doesn't sound like we need dive deeper into the proposed solution anyway.